lib/Model/Type.cpp

```diff
@@ -443,7 +443,7 @@ static VoidConstResult isVoidConst(const QualifiedType *QualType) {
       // We know that it's const-qualified here, and it only has one
       // qualifier, hence we can skip the const-qualifier.
       Result.IsConst = true;
-      if (not QualType->UnqualifiedType.Root)
+      if (QualType->UnqualifiedType.hasNullRoot())
         return Result;
```

Comment:
Can we do something like `.get() == nullptr`?

Reply:
Eh, I'd love to, but the previous implementation of `get()` asserts that `Root` is not null. So we'd get an error on `get()` here.

Reply:
I could change `get()` not to assert, though.

Reply:
Mmmh, I'm not sure why here we're checking if `Root` is `nullptr`.
`Root` is `nullptr` only during deserialization, in all other cases it's not a valid state.
I think we can just drop this check, it's probably a leftover of a darker past.

Reply:
Could be. I'm going to drop this, and make `hasNullRoot` private then.

Reply:
Done

Reply:
> there are valid situations where we need to have some `QualifiedTypes` with null roots for a while

What are those? `TupleTreeReference`s were not designed for this.

Reply:
Since the type system can have loops (consider pointers), the type creation pass does the following: first, it generates all types except pointers, then it updates all the fields corresponding to pointer with the right type. Between the two phases, the pointer fields exist in the model, but they do not have a valid `UnqualifiedType` yet, and I want to check that that is the case. In some cases, an additional wrapper might have been generated for the pointer, so I need to descend the wrapper structures until I find one with an empty `UnqualifiedType`, but that is an implementation detail that can be worked around.

Reply:
TL;DR I want to check if a field of a struct/union has a type that has yet to be decided, so I can use whatever mechanism is exposed at a higher level (e.g. maybe check if the Type is Invalid?)

Reply:
Sounds like you can use `TupleTreeReference::isValid`.
Having a valid path but not a valid root (which is what we're discussing here) makes sense only if upon creation of the `TupleTreeRereference` the root (`model::Binary`) has not been instantiated yet.
But this doesn't seem to be the case.

Reply:
Agreed

Reply:
I tried this with `isValid` and it works. Thanks @AlviseDeFaveri for the clarification. I'm going to close this conversation.